### ISSUE TITLE
docs: two more container-leak fixes — timed-out docker exec now kills its container, `run_on=` reclaims its instance on shutdown (PR #4109)

### ISSUE THREAD

**Issue by MervinPraison** [MervinPraison/PraisonAIDocs#3104]
Upstream PR: [MervinPraison/PraisonAI#4109](https://github.com/MervinPraison/PraisonAI/pull/4109) (merged 2026-08-20, head SHA `2cffd51012733f04dc3379ece65b3720ddbe5e2d`).

PR #4109 lands **two pre-existing container-leak fixes** that are direct siblings of the shutdown-leak work in PR #4107. Each one changes documented behaviour (or invalidates a doc claim we already published under the PR #4107 issue #3099). The docs currently describe the pre-fix world in two places and need to be brought in line with the code that shipped to `main`.

This is a **docs-only** issue — the code is already merged. All updates below go in `docs/features/` (never `docs/concepts/`, per `AGENTS.md` §1.8).

---

## Executive summary of what changed in the SDK

| # | Severity | Area | Change |
|---|----------|------|--------|
| 1 | **HIGH** | Docker sandbox `execute()` timeout | Every `docker run` now gets `--name praisonai-<execution_id>` and `--label praisonai=sandbox-exec`. On timeout, the sandbox issues `docker kill <name>` before `proc.kill()` — previously `proc.kill()` ended only the docker **client** and the container it started kept running, with a random Docker-assigned name that neither the timeout handler nor `praisonai managed ps` could find. |
| 2 | **HIGH** | `ComputeManagedAgent` (`run_on=`) shutdown | The backend now registers a `weakref.finalize` at provision time that reclaims the instance when the backend is collected **and** at interpreter exit. Previously there was no finalizer at all, so an instance survived every script — a Docker container per run, or a billed cloud instance for `e2b`/`modal`/`daytona`/`tenki`/`flyio` whose only other reaper is the provider's idle timer (and **`docker` and `flyio` have none**). |
| 3 | supporting | Explicit `ashutdown()` | The finalizer is detached on explicit shutdown so an intentional teardown is never double-reclaimed. |
| 4 | supporting | Event-loop-safe release | The finalizer body (`_release()`) bridges through a `ThreadPoolExecutor` when GC lands on a thread that already drives a loop, instead of calling `asyncio.run()` directly (which would `RuntimeError` and silently leak). |
| 5 | supporting | Ephemeral vs managed labelling | The new `execute()`-path containers are labelled `praisonai=sandbox-exec`, **deliberately NOT** `praisonai=managed`. `praisonai=managed` is the sole input to `DockerCompute.list_instances()`; tagging ephemeral `--rm` containers as managed would make `praisonai managed ps` list a name (`praisonai-<uuid>`) that `managed stop` (which resolves `praisonai_<id>`) could never reclaim. |

There is also a signature-compat fix to the docker-py test double: `Container.commit(..., changes=..., **kwargs)` now mirrors the real docker-py API so the secret-scrubbing capture path from PR #4107 still commits successfully in tests.

### How the two leaks relate

Both PRs fight the same class of bug — *a container that outlived the thing that created it* — but at different layers:

- **PR #4107** fixed the leak in `SyncComputeProvider._offload()` and `_shutdown_orphan()`: the release ran during atexit, `run_in_executor` raised, the error was swallowed, and Docker/Fly.io instances lived indefinitely.
- **PR #4109** fixes the two remaining leaks the PR #4107 review surfaced:
  1. There was no finalizer to invoke that path in the first place on `ComputeManagedAgent`, so `run_on=` instances never got the chance to be reclaimed.
  2. The `SandboxManager` docker `execute()` path had a separate leak of its own — killing the client, not the container.

Together, `#4107 + #4109` close the container-leak story end-to-end for both `sandbox=` and `run_on=`.

---

### A. `docs/features/sandbox.mdx` — Timeout Behavior & Resource Cleanup table (~L780)

**Currently says:** *"docker: Client process killed **and** `docker kill <container>` issued on the named container (`praisonai-<execution_id>`) — Container stopped — resource limits enforced"*.

This wording was written **anticipating** PR #4109's fix (it appears alongside a sequence diagram that already shows the correct behaviour). The wording is now correct — but the corresponding narrative earlier on the page (the sequence diagram introduction, and the "Why containers are no longer orphaned" paragraph) should be tied explicitly to PR #4109 and marked as the shipped behaviour, not aspirational.

**Update needed:**
- Add a `<Note>` (or reuse the existing "Why containers are no longer orphaned" paragraph — verify it is present at ~L840) explicitly citing [PR #4109](https://github.com/MervinPraison/PraisonAI/pull/4109) as the fix, so readers on the changelog can find the source.
- Clarify that the fix has two moving parts:
  1. Every `docker run` from `execute()` gets `--name praisonai-<execution_id>` (previously unnamed → random Docker name).
  2. On `asyncio.TimeoutError`, the sandbox spawns `docker kill <container_name>` and awaits it **before** calling `proc.kill()` — killing the container, not just the client.
- Add a short "before" reproduction so the risk is legible:
  ```python
  cfg = SandboxConfig.docker()
  cfg.resource_limits.timeout_seconds = 5
  agent.execute_code_sync("import time; time.sleep(300)", run_in=cfg)
  # -> SandboxStatus.TIMEOUT
  # docker ps  # (pre-#4109) still shows a running container with a random name
  # praisonai managed ps  # can't see it — no managed label, no known name
  ```
- Note the deliberate labelling choice: `execute()` containers carry `praisonai=sandbox-exec`, **not** `praisonai=managed`. The two labels have different lifecycles (ephemeral `--rm` per execution vs. long-lived managed instance) and different name shapes (`praisonai-<uuid>` vs. `praisonai_<id>`). `praisonai managed ps` deliberately does not list them; use `docker ps --filter label=praisonai=sandbox-exec` if you need to see them.

### B. `docs/features/sandbox-backends.mdx` — Docker section

If the page names `praisonai=managed` as *the* praisonai label users can filter on, add the second label (`praisonai=sandbox-exec`) and explain the split. Two labels, two lifecycles.

Suggested table:

| Label | Applied by | Container lifetime | Listed by `praisonai managed ps` |
|:--|:--|:--|:--|
| `praisonai=managed` | `DockerCompute` (managed instance for `run_on=` / `compute=`) | Long-lived; reclaimed by `managed stop`, finalizer, or `ashutdown()` | ✅ |
| `praisonai=sandbox-exec` | `DockerSandbox.execute()` (per-execution) | Ephemeral (`--rm`); killed on timeout by name | ❌ |

### C. `docs/features/reclaim-stray-sandboxes.mdx` — the "silent leak on shutdown" section

The current section documents PR #4107's fix. It needs a second paragraph documenting the PR #4109 fix that sits **upstream** of it — the leak PR #4107 patched only fired when a finalizer actually ran, and on `ComputeManagedAgent` no finalizer had ever been registered.

**Update needed:**
- Add a subsection **"…and the `run_on=` finalizer that was never registered ([PR #4109](https://github.com/MervinPraison/PraisonAI/pull/4109))"**.
- Explain the shape of the pre-#4109 leak:
  - `ComputeManagedAgent` defaults to `keep_alive=True` — the instance is *supposed* to survive between calls, not between processes.
  - Without a finalizer, the instance survived the whole process. For `docker` that was a container per script; for `e2b`, `modal`, `daytona`, `tenki`, `flyio` it was a billed cloud instance whose only other reaper is the provider's idle timer.
  - **Docker and Fly.io have no idle timer**, so those instances lived indefinitely.
- Document the new mechanism:
  - `weakref.finalize(self, _release, provider, self._instance, self._place)` is registered inside `_ensure()` right after `provision()` succeeds.
  - The finalizer runs on garbage collection **and** at interpreter exit.
  - It deliberately does not close over the backend `self` — closing over it would keep the backend alive and prevent the very collection it waits for.
- Document `_release()` — a module-level function in `praisonai/integrations/compute_managed_agent.py`:
  - Detects an already-running event loop via `asyncio.get_running_loop()`.
  - If none: `asyncio.run(_shutdown())`.
  - If one: bridges through `concurrent.futures.ThreadPoolExecutor(max_workers=1)` so the shutdown coroutine runs on its own loop instead of `RuntimeError`-ing. This matches the pattern `SharedCompute` already uses.
  - Errors are logged at warning and swallowed — teardown stays quiet.
- Document the explicit-shutdown handshake: `ashutdown()` calls `self._finalizer.detach()` before shutting down, so a deliberate teardown is never double-reclaimed by the finalizer that fires shortly after.
- Cross-link the PR #4107 subsection: "PR #4107 made the release path survive interpreter shutdown; PR #4109 made sure the release path is actually invoked on `run_on=`."

Recommended reader guidance for older-version cleanup remains the same: `praisonai managed ps` / `managed stop` (and `docker container prune`).

### D. `docs/features/placement.mdx` — the PR #4107 note (~L128)

The note already says: *"PR #4107 finishes the job by making `DockerCompute` inherit `SyncComputeProvider`…"* — extend it with one sentence:

*"PR [#4109](https://github.com/MervinPraison/PraisonAI/pull/4109) closes two remaining leaks in the same story: the timed-out docker `execute()` container is now killed by name, and `ComputeManagedAgent` finally registers a `weakref.finalize` so `run_on=` instances are reclaimed when the backend is collected or the process exits."*

### E. `docs/features/run-on-docker.mdx` — Best Practices accordion

The current page has: *"**Call `.shutdown()` when you disable `keep_alive`** — With `keep_alive=False`, tear the container down with `.shutdown()` or use the backend as a context manager…"*.

This is now incomplete — with the finalizer, `keep_alive=True` is also cleaned up (on GC / interpreter exit), and explicit `.shutdown()` remains the right call whenever you want teardown *now* rather than *eventually*.

**Update needed:**
- Add a second accordion or a paragraph inside the existing one: **"With `keep_alive=True`, the container is still cleaned up automatically."** Explain the finalizer briefly (one line: "a `weakref.finalize` on the backend reclaims the container when the backend is collected or the process exits"), and note that `.shutdown()` still exists — call it when you need teardown to happen *now* rather than *whenever GC runs*.
- Optionally add a `<Note>` linking to `reclaim-stray-sandboxes.mdx#…-and-the-run-on-finalizer-that-was-never-registered` for readers on an older version.

### F. `docs/features/sandbox-guarantees.mdx` — reliability guarantee

If the page has a "Reliability" section or a "what does the sandbox promise" matrix, add a row:

| Guarantee | How it is achieved |
|:--|:--|
| A timed-out docker execution leaves no running container | Every `docker run` is named `praisonai-<execution_id>`; on timeout the sandbox issues `docker kill <name>` before killing the client (PR [#4109](https://github.com/MervinPraison/PraisonAI/pull/4109)). |
| A `run_on=` instance is reclaimed at process exit | `ComputeManagedAgent` registers a `weakref.finalize` at provision that runs on GC and at interpreter exit (PR [#4109](https://github.com/MervinPraison/PraisonAI/pull/4109)); explicit `ashutdown()` detaches the finalizer to avoid double-release. |

### G. Optional new page — `docs/features/sandbox-container-lifecycle.mdx`

Nice-to-have single-page story that stitches PR #4107 and PR #4109 together for a reader who lands on the changelog. It could sit next to `reclaim-stray-sandboxes.mdx` in the sidebar and give the whole picture in one place:
- What each label means (`praisonai=managed` vs `praisonai=sandbox-exec`).
- How the finalizer + `_offload()` interpreter-shutdown fallback compose end to end.
- The exact repro that used to leak, and what to look for in `docker ps` today.

Not required for this issue — the affected pages above cover the same ground when read together.

---

## SDK ground truth (paths, symbols, defaults)

All paths in PR #4109 live under `src/praisonai-sandbox/` and `src/praisonai/` in the PraisonAI repo. When documenting, treat these as source of truth (follow `AGENTS.md` §1.2 — read the file, do not guess):

| Symbol / constant | Location | What it is |
|---|---|---|
| `container_name = f"praisonai-{execution_id}"` | `praisonai-sandbox/praisonai_sandbox/docker.py` (in `execute()`) | Deterministic per-execution container name. Passed as `--name` to `docker run` and to `docker kill` on timeout. |
| `"--label", "praisonai=sandbox-exec"` | same file, `_build_docker_command()` | New label on `execute()` containers. **Deliberately NOT** `praisonai=managed`. |
| Timeout kill sequence | same file, `execute()` timeout branch | `await create_subprocess_exec("docker", "kill", container_name).wait()` then `proc.kill()` then `proc.wait()`. Exceptions during kill are logged at `warning` and do not mask the `SandboxStatus.TIMEOUT` return. |
| `_build_docker_command(..., container_name: str = "")` | same file | Signature grew a new `container_name` argument. `run_command()` still passes `""` today (its own path does not need it); `execute()` passes the deterministic name. |
| `ComputeManagedAgent._finalizer` | `praisonai/praisonai/integrations/compute_managed_agent.py` | New instance attribute (`None` at `__init__`, set inside `_ensure()`). |
| `weakref.finalize(self, _release, provider, self._instance, self._place)` | same file, inside `_ensure()` | Registers `_release` to run when the backend is collected or the process exits. Deliberately does not close over `self`. |
| `_release(provider, instance_id, place)` | same file, module level | Loop-safe shutdown: `asyncio.get_running_loop()` → if it raises, `asyncio.run(_shutdown())`; if a loop is already running, bridge through `ThreadPoolExecutor(max_workers=1).submit(asyncio.run, _shutdown()).result()`. Warnings only; exceptions swallowed. |
| `ashutdown()` finalizer detach | same file | Now begins with `if self._finalizer is not None: self._finalizer.detach(); self._finalizer = None` so an explicit shutdown never double-releases. |
| Test-double `Container.commit(..., changes=None, **kwargs)` | `praisonai/tests/unit/compute/test_docker_capture.py` | Signature mirrors docker-py; presence of `changes` (used by PR #4107's secret scrubbing) must not turn a successful commit into a failure in the fake. |

Tests that pin the new behaviour (name them in any doc example that references the guarantees):

- `test_a_hosted_backend_reclaims_its_instance_when_collected` — `run_on=` finalizer.
- `test_an_explicit_shutdown_cancels_the_finalizer` — no double-release on `ashutdown()`.
- `test_release_reclaims_even_on_an_event_loop_thread` — GC on a live loop still releases.
- `test_a_timed_out_execution_does_not_leave_a_container` — docker `execute()` timeout kills the container.

Verified numbers from the PR body (useful for the reliability page):

| | Before | After |
|---|---|---|
| Containers after a timeout | 1 | **0** |
| Containers after a `run_on=` script exits | 1 | **0** |
| Ordinary run | 42 | 42 |

143 passing in `praisonai-sandbox`, 187 across the placement suites.

---

## Files to update (concrete list)

1. `docs/features/sandbox.mdx` — the Timeout Behavior & Resource Cleanup subsection (~L780) and the "Why containers are no longer orphaned" paragraph (~L840): cite PR #4109 as the fix, add the "before" repro, explain the two-label split.
2. `docs/features/sandbox-backends.mdx` — add the two-label table (`praisonai=managed` vs `praisonai=sandbox-exec`) to the Docker section.
3. `docs/features/reclaim-stray-sandboxes.mdx` — new subsection **"…and the `run_on=` finalizer that was never registered ([PR #4109](https://github.com/MervinPraison/PraisonAI/pull/4109))"**, immediately after the existing PR #4107 subsection.
4. `docs/features/placement.mdx` — extend the existing PR #4107 note (~L128) with one sentence citing PR #4109 for the two remaining leaks.
5. `docs/features/run-on-docker.mdx` — augment the "Call `.shutdown()` when you disable `keep_alive`" accordion with the `keep_alive=True` cleanup story via the new finalizer.
6. `docs/features/sandbox-guarantees.mdx` — add the two guarantee rows above.

Optional (nice-to-have): new page `docs/features/sandbox-container-lifecycle.mdx` that stitches PR #4107 and PR #4109 into one story. If added, wire it into `docs.json` under the **Features** group only (never Concepts).

---

## Style requirements (from `AGENTS.md`)

- **Placement:** all changes go under `docs/features/` — never `docs/concepts/` (that folder is human-only per §1.8).
- **Agent-centric first:** every code sample should lead with `Agent(...)` (or `SandboxConfig` on the sandbox pages), not with `DockerSandbox` internals or `ComputeManagedAgent` internals. Show the fix from the user's perspective (*"your timed-out execution no longer leaves a container behind"*, *"your `run_on=` instance is reclaimed when your script exits"*) before dropping to the mechanism.
- **Progressive disclosure:** short first, deep second. Beginners should see *"this is fixed, use as before"* up top; the `weakref.finalize` walk and the ThreadPoolExecutor bridge are subsections.
- **Mermaid diagrams** for any new page, standard colour scheme (`#8B0000`, `#189AB4`, `#10B981`, `#F59E0B`, `#6366F1` on `#7C90A0` strokes, white text). If updating the timeout sequence diagram already on `sandbox.mdx`, keep it — it already shows the correct flow.
- **Mintlify components:** `<Steps>` in Quick Start, `<AccordionGroup>` in Best Practices, `<CardGroup>` in Related, `<Warning>` for "before the fix" callouts.
- **Import style:** friendly, top-level — `from praisonaiagents import Agent, SandboxConfig`. Reach into `praisonai_sandbox.docker` or `praisonai.integrations.compute_managed_agent` **only** when documenting the internal symbols in the ground-truth table.
- Cross-link every page you touch to `reclaim-stray-sandboxes.mdx` and `placement.mdx` — the container-lifecycle story spans several pages and readers need the map.
- No forbidden phrases (*"In this section…"*, *"As you can see…"*, *"It's important to note…"*, *"Please note…"*, *"Let's take a look at…"*, *"The following example shows…"*).

---

## Verification checklist for the doc-writing agent

- [ ] Every code sample compiles and would run against merged `main` (head SHA `2cffd51012733f04dc3379ece65b3720ddbe5e2d` or later).
- [ ] All symbol names, constants, and default values match the SDK exactly (see ground-truth table above; re-open the file to confirm — never guess).
- [ ] `docs.json` updated only under the **Features** group if a new page is added — never under **Concepts**, never under the auto-managed `docs/js/` or `docs/rust/` groups.
- [ ] Every affected page still has: frontmatter, hero diagram (if applicable), `<Steps>` Quick Start, Configuration Options link, `<AccordionGroup>` Best Practices, `<CardGroup>` Related.
- [ ] The `execute()`-path container label is documented as `praisonai=sandbox-exec` (not `praisonai=managed`) everywhere it appears; the reason (`managed ps` lookup shape mismatch) is stated at least once.
- [ ] The `run_on=` finalizer is documented as running on **both** garbage collection **and** interpreter exit — not "one or the other".
- [ ] The event-loop-safe release path (`ThreadPoolExecutor` bridge) is mentioned wherever `weakref.finalize` is described, so readers understand why the naive `asyncio.run()` version would have leaked from inside async workflow code.
- [ ] Explicit `ashutdown()` is documented as detaching the finalizer (`self._finalizer.detach()`) to avoid double-release.
- [ ] The two-container table row *"Containers after a timeout: 1 → 0 / Containers after a `run_on=` script exits: 1 → 0 / Ordinary run: 42 → 42"* is referenced somewhere as the verified before/after so the guarantee is not vague.
